server/pulp/server/controllers/content.py

```diff
@@ -128,113 +137,392 @@ def refresh_content_source(content_source_id=None):
-        [catalog_entry],
-        tags=tags)
+
+    def __init__(self, task_id):
+        """
+        :param task_id: identifies the task being performed
+        :type  task_id: str
+        """
+        StatusMixin.__init__(self, task_id, LazyStatusConduitException)
+
+    def __str__(self):
+        return 'LazyStatusConduit'
+
+
+class LazyUnitDownloadStep(Step, DownloadEventListener):
+    """
+    A Step that downloads all the given requests. The downloader is configured
+    to download from the Pulp Streamer components.
+
+    :ivar download_requests: The download requests the step will process.
+    :type download_requests: list of nectar.request.DownloadRequest
+    :ivar download_config:   The keyword args used to initialize the Nectar
+                             downloader configuration.
+    :type download_config:   dict
+    :ivar downloader:        The Nectar downloader used to fetch the requests.
+    :type downloader:        nectar.downloaders.threaded.HTTPThreadedDownloader
+    """
+
+    def __init__(self, step_type, step_description, lazy_status_conduit, download_requests):
+        """
+        Initializes a Step that downloads all the download requests provided.
+
+        :param lazy_status_conduit: Conduit used to update the task status.
+        :type  lazy_status_conduit: LazyStatusConduit
+        :param download_requests:   List of download requests to process.
+        :type  download_requests:   list of nectar.request.DownloadRequest
+        """
+        super(LazyUnitDownloadStep, self).__init__(
+            step_type=step_type,
+            status_conduit=lazy_status_conduit,
+        )
+        self.description = step_description
+        self.download_requests = download_requests
+        self.download_config = {
+            MAX_CONCURRENT: int(pulp_conf.get('lazy', 'download_concurrency')),
+            HEADERS: {PULP_STREAM_REQUEST_HEADER: 'true'},
+            SSL_VALIDATION: True
+        }
+        self.downloader = HTTPThreadedDownloader(
+            DownloaderConfig(**self.download_config),
+            self
+        )
+
+    def _process_block(self, item=None):
+        """
+        This block is called by the `process` loop. This is overridden because
+        success and failure is determined during the EventListener callbacks,
+        which will handle updating the progress. Since `item` is not used, this
+        does not make use of `process_main` and simply calls the downloader.
+
+        Inherited from Step.
+
+        :param item: Unused.
+        :type  item: None
+        """
+        self.downloader.download(self.download_requests)
+
+    def get_total(self):
+        """
+        The total number of download requests so progress reporting occurs at
+        the file level.
+
+        Inherited from Step.
+
+        :return: The number of download requests this step will process.
+        :rtype:  int
+        """
+        return len(self.download_requests)
```

Comment:
I see that this will be difficult if ``self.download_requests`` is a generator. But the memory implications alone of this not being a generator are I think worth the cost of not reporting a denominator in the progress. I wonder if anything would complain if this returned 0 or -1? Or ``None``?

Otherwise, I suggest storing the number of files on the ``DeferredDownload`` model, allowing you to quickly add them up with a relatively inexpensive DB call.